File: truevfs-ext/truevfs-ext-pace/src/main/java/net/java/truevfs/ext/pace/PaceMediator.java
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.ext.pace;

import javax.annotation.concurrent.ThreadSafe;
import net.java.truevfs.comp.inst.InstrumentingManager;
import net.java.truevfs.comp.jmx.JmxMediator;
import net.java.truevfs.kernel.spec.FsController;
import net.java.truevfs.kernel.spec.FsManager;

/**
 * A mediator for the instrumentation of the TrueVFS Kernel with a
 * {@linkplain PaceManager pace manager}.
 *
 * @author Christian Schlichtherle
 */
@ThreadSafe
final class PaceMediator extends JmxMediator<PaceMediator> {

    static final PaceMediator SINGLETON = new PaceMediator();

    private PaceMediator() { }

    @Override
    public FsManager instrument(FsManager object) {
        return start(new PaceManager(this, object));
    }

    @Override
    public FsController instrument(
            InstrumentingManager<PaceMediator> origin,
            FsController object) {
        return new PaceController((PaceManager) origin, object);
    }
}

File: truevfs-ext/truevfs-ext-pacemanager/src/main/scala/net/java/truevfs/ext/pacemanager/PaceMediator.scala
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.ext.pacemanager

import javax.annotation.concurrent.ThreadSafe
import net.java.truevfs.comp.inst.InstrumentingManager
import net.java.truevfs.comp.jmx.JmxMediator
import net.java.truevfs.kernel.spec.FsController
import net.java.truevfs.kernel.spec.FsManager

/**
  * A mediator for the instrumentation of the TrueVFS Kernel with a
  * [[net.java.truevfs.ext.pacemanager.PaceManager]].
  *
  * @author Christian Schlichtherle
  */
@ThreadSafe
private class PaceMediator extends JmxMediator[PaceMediator] {

  override def instrument(obj: FsManager) = start(new PaceManager(this, obj))

  override def instrument(origin: InstrumentingManager[PaceMediator], obj: FsController) =
    new PaceController(origin.asInstanceOf[PaceManager], obj)
}

private object PaceMediator extends PaceMediator
